libs/utils/trace.py

```diff
@@ -674,6 +677,23 @@ def _sanitize_CpuFrequency(self):
             return
         df = self._dfg_trace_event('cpu_frequency')
         clusters = self.platform['clusters']
+
+        # devlib introduces fake cpu_frequency events to make sure the
+        # frequency is known at the "beginning" of a trace
+        if self.hasEvents('cpu_frequency_devlib'):
+            devlib_freq = self._dfg_trace_event('cpu_frequency_devlib')
+            devlib_freq.rename(columns={'cpu_id':'cpu'}, inplace=True)
+            devlib_freq.rename(columns={'state':'frequency'}, inplace=True)
+            # Only use fake events if introduced before any cpufreq governor
+            # did, otherwise they may end up being interleaved with real events
+            # and make the the frequency coherency check fail
+            if len(devlib_freq) > 0:
+                first_devlib_freq = devlib_freq.iloc[:self.platform['cpus_count']]
+                if df.index[0] > first_devlib_freq.index[-1]:
+                    df = pd.concat([first_devlib_freq, df])
+            setattr(self.ftrace.cpu_frequency, 'data_frame', df)
```

Comment:
In general I prefer this alterations to happen in trappy. Most of the `_sanitize_*` in `Trace()` should be moved to trappy.

However, in this particular case it's operating on stuff coming from devlib, so I'd rather leave it in `lisa`.